rosbag2_compression/src/rosbag2_compression/compression_factory_impl.hpp

```diff
@@ -20,67 +20,116 @@
 #include <stdexcept>
 #include <string>
 
+#include "pluginlib/class_loader.hpp"
+
 #include "logging.hpp"
 #include "rosbag2_compression/compression_factory.hpp"
-#include "rosbag2_compression/zstd_compressor.hpp"
-#include "rosbag2_compression/zstd_decompressor.hpp"
 
-namespace
+namespace rosbag2_compression
 {
 
-constexpr const char kCompressionFormatZstd[] = "zstd";
+using rosbag2_compression::BaseCompressorInterface;
+using rosbag2_compression::BaseDecompressorInterface;
+
+template<typename T>
+struct CompressionTraits
+{};
 
-/// Verify whether two case-insensitive chars are equal
-bool compare_char(const char c1, const char c2)
+template<>
+struct CompressionTraits<BaseCompressorInterface>
 {
-  return c1 == c2 || std::tolower(c1) == std::tolower(c2);
-}
+  static constexpr const char * name = "rosbag2_compression::BaseCompressorInterface";
+};
 
-/// Performs a case-insensitive comparison of two strings
-bool case_insensitive_compare(const std::string & str1, const std::string & str2) noexcept
+template<>
+struct CompressionTraits<BaseDecompressorInterface>
 {
-  return (str1.size() == str2.size()) &&
-         std::equal(str1.begin(), str1.end(), str2.begin(), &compare_char);
+  static constexpr const char * name = "rosbag2_compression::BaseDecompressorInterface";
+};
+
+template<typename InterfaceT>
+std::shared_ptr<pluginlib::ClassLoader<InterfaceT>>
+get_class_loader()
+{
+  const char * lookup_name = CompressionTraits<InterfaceT>::name;
```

Comment:
Can probably be `const auto` or maybe `std::string` since `ClassLoader` ctor is implicitly converting to a std::string.